File: behavioral-design-pattern/memento-pattern/memento.py
"""
Memento Pattern
================
Category: Behavioral Design Pattern

Intent:
    Without violating encapsulation, capture and externalize an object's
    internal state so that the object can be restored to this state later.

When to use:
    - You need to save and restore snapshots of an object's state
    - A direct interface to obtaining the state would expose implementation
      details and break encapsulation
    - You need undo/rollback functionality

Key Participants:
    - Originator (GameCharacter): the object whose state needs saving;
      creates a memento containing a snapshot of its current internal state
    - Memento (GameSave): stores the originator's internal state;
      protects against access by objects other than the originator
    - Caretaker (SaveManager): responsible for keeping the memento safe;
      never operates on or examines the contents of a memento
"""

from __future__ import annotations
from dataclasses import dataclass, field
from datetime import datetime


# --- Memento ---

@dataclass(frozen=True)
class GameSave:
    """
    Memento: an immutable snapshot of the character's state.
    frozen=True prevents anyone from tampering with the saved state.
    """
    health: int
    level: int
    position_x: float
    position_y: float
    inventory: tuple[str, ...]  # immutable copy of inventory
    timestamp: str
    label: str

    def __str__(self) -> str:
        return (
            f"[{self.timestamp}] '{self.label}' -- "
            f"HP:{self.health} Lv:{self.level} "
            f"Pos:({self.position_x},{self.position_y}) "
            f"Items:{list(self.inventory)}"
        )


# --- Originator ---

class GameCharacter:
    """
    Originator: the game character whose state we want to save and restore.
    Only the character itself knows how to create and restore from a memento.
    """

    def __init__(self, name: str) -> None:
        self.name = name
        self.health = 100
        self.level = 1
        self.position_x = 0.0
        self.position_y = 0.0
        self.inventory: list[str] = []

    def take_damage(self, amount: int) -> None:
        self.health = max(0, self.health - amount)
        print(f"  {self.name} took {amount} damage. HP: {self.health}")

    def level_up(self) -> None:
        self.level += 1
        self.health = 100  # full heal on level up
        print(f"  {self.name} leveled up to Lv.{self.level}! HP restored.")

    def move_to(self, x: float, y: float) -> None:
        self.position_x = x
        self.position_y = y
        print(f"  {self.name} moved to ({x}, {y})")

    def pick_up(self, item: str) -> None:
        self.inventory.append(item)
        print(f"  {self.name} picked up '{item}'")

    # --- Memento creation and restoration ---

    def save(self, label: str) -> GameSave:
        """Create a memento capturing the current state."""
        memento = GameSave(
            health=self.health,
            level=self.level,
            position_x=self.position_x,
            position_y=self.position_y,
            inventory=tuple(self.inventory),  # snapshot as immutable tuple
            timestamp=datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            label=label,
        )
        print(f"  >> State saved: '{label}'")
        return memento

    def restore(self, memento: GameSave) -> None:
        """Restore state from a memento."""
        self.health = memento.health
        self.level = memento.level
        self.position_x = memento.position_x
        self.position_y = memento.position_y
        self.inventory = list(memento.inventory)  # mutable copy from tuple
        print(f"  >> State restored from '{memento.label}'")

    def status(self) -> str:
        return (
            f"  {self.name}: HP={self.health} Lv={self.level} "
            f"Pos=({self.position_x},{self.position_y}) "
            f"Inventory={self.inventory}"
        )


# --- Caretaker ---

class SaveManager:
    """
    Caretaker: manages a collection of saves without knowing their internals.
    It never inspects or modifies the memento's state.
    """

    def __init__(self) -> None:
        self._saves: dict[str, GameSave] = {}

    def store(self, key: str, save: GameSave) -> None:
        """Store a save under a named slot."""
        self._saves[key] = save

    def load(self, key: str) -> GameSave | None:
        """Retrieve a save by slot name."""
        save = self._saves.get(key)
        if save is None:
            print(f"  >> No save found for slot '{key}'")
        return save

    def list_saves(self) -> None:
        """List all stored saves."""
        if not self._saves:
            print("  No saves stored.")
            return
        print("  Stored saves:")
        for key, save in self._saves.items():
            print(f"    [{key}] {save}")


# --- Main ---

def main() -> None:
    hero = GameCharacter("Warrior")
    manager = SaveManager()

    print("=== Starting the adventure ===")
    print(hero.status())
    print()

    # Play through some actions
    hero.move_to(10.0, 20.0)
    hero.pick_up("Iron Sword")
    hero.level_up()
    print(hero.status())
    print()

    # Save at checkpoint 1
    checkpoint1 = hero.save("Before the dungeon")
    manager.store("checkpoint1", checkpoint1)
    print()

    # Enter the dungeon -- things go badly
    hero.move_to(50.0, 80.0)
    hero.take_damage(60)
    hero.pick_up("Rusty Key")
    hero.take_damage(30)
    print(hero.status())
    print()

    # Save the risky state too
    checkpoint2 = hero.save("Deep in dungeon (low HP)")
    manager.store("checkpoint2", checkpoint2)
    print()

    # Even worse...
    hero.take_damage(50)
    print(hero.status())
    print()

    # List all saves
    print("=== Reviewing saves ===")
    manager.list_saves()
    print()

    # Restore to the safe checkpoint
    print("=== Restoring to checkpoint 1 ===")
    save = manager.load("checkpoint1")
    if save:
        hero.restore(save)
    print(hero.status())


if __name__ == "__main__":
    main()

File: behavioral-design-pattern/memento-pattern/memento.ts
/**
 * Memento Pattern
 * ================
 * Category: Behavioral Design Pattern
 *
 * Intent:
 *   Without violating encapsulation, capture and externalize an object's
 *   internal state so that the object can be restored to this state later.
 *
 * When to use:
 *   - You need to save and restore snapshots of an object's state
 *   - A direct interface to obtaining the state would expose implementation
 *     details and break encapsulation
 *   - You need undo/rollback functionality
 *
 * Key Participants:
 *   - Originator (GameCharacter): the object whose state needs saving;
 *     creates a memento containing a snapshot of its current internal state
 *   - Memento (GameSave): stores the originator's internal state;
 *     protects against access by objects other than the originator
 *   - Caretaker (SaveManager): responsible for keeping the memento safe;
 *     never operates on or examines the contents of a memento
 */

// --- Memento ---

/**
 * GameSave is the memento: an immutable snapshot of the character's state.
 * We use Readonly<> to prevent external mutation.
 */
interface GameSaveData {
  readonly health: number;
  readonly level: number;
  readonly positionX: number;
  readonly positionY: number;
  readonly inventory: readonly string[]; // immutable copy
  readonly timestamp: string;
  readonly label: string;
}

class GameSave implements GameSaveData {
  readonly health: number;
  readonly level: number;
  readonly positionX: number;
  readonly positionY: number;
  readonly inventory: readonly string[];
  readonly timestamp: string;
  readonly label: string;

  constructor(data: GameSaveData) {
    this.health = data.health;
    this.level = data.level;
    this.positionX = data.positionX;
    this.positionY = data.positionY;
    this.inventory = Object.freeze([...data.inventory]); // deep-freeze the copy
    this.timestamp = data.timestamp;
    this.label = data.label;
  }

  toString(): string {
    return (
      `[${this.timestamp}] '${this.label}' -- ` +
      `HP:${this.health} Lv:${this.level} ` +
      `Pos:(${this.positionX},${this.positionY}) ` +
      `Items:[${this.inventory.join(", ")}]`
    );
  }
}

// --- Originator ---

/**
 * GameCharacter is the originator whose state we want to save and restore.
 * Only the character itself knows how to create and restore from a memento.
 */
class GameCharacter {
  health = 100;
  level = 1;
  positionX = 0;
  positionY = 0;
  inventory: string[] = [];

  constructor(public readonly name: string) {}

  takeDamage(amount: number): void {
    this.health = Math.max(0, this.health - amount);
    console.log(`  ${this.name} took ${amount} damage. HP: ${this.health}`);
  }

  levelUp(): void {
    this.level += 1;
    this.health = 100; // full heal on level up
    console.log(`  ${this.name} leveled up to Lv.${this.level}! HP restored.`);
  }

  moveTo(x: number, y: number): void {
    this.positionX = x;
    this.positionY = y;
    console.log(`  ${this.name} moved to (${x}, ${y})`);
  }

  pickUp(item: string): void {
    this.inventory.push(item);
    console.log(`  ${this.name} picked up '${item}'`);
  }

  // --- Memento creation and restoration ---

  /** Create a memento capturing the current state. */
  save(label: string): GameSave {
    const memento = new GameSave({
      health: this.health,
      level: this.level,
      positionX: this.positionX,
      positionY: this.positionY,
      inventory: [...this.inventory], // snapshot copy
      timestamp: new Date().toISOString().replace("T", " ").slice(0, 19),
      label,
    });
    console.log(`  >> State saved: '${label}'`);
    return memento;
  }

  /** Restore state from a memento. */
  restore(memento: GameSave): void {
    this.health = memento.health;
    this.level = memento.level;
    this.positionX = memento.positionX;
    this.positionY = memento.positionY;
    this.inventory = [...memento.inventory]; // mutable copy from frozen array
    console.log(`  >> State restored from '${memento.label}'`);
  }

  status(): string {
    return (
      `  ${this.name}: HP=${this.health} Lv=${this.level} ` +
      `Pos=(${this.positionX},${this.positionY}) ` +
      `Inventory=[${this.inventory.join(", ")}]`
    );
  }
}

// --- Caretaker ---

/**
 * SaveManager is the caretaker: it manages saves without knowing their internals.
 * It never inspects or modifies the memento's state.
 */
class SaveManager {
  private saves = new Map<string, GameSave>();

  store(key: string, save: GameSave): void {
    this.saves.set(key, save);
  }

  load(key: string): GameSave | undefined {
    const save = this.saves.get(key);
    if (!save) {
      console.log(`  >> No save found for slot '${key}'`);
    }
    return save;
  }

  listSaves(): void {
    if (this.saves.size === 0) {
      console.log("  No saves stored.");
      return;
    }
    console.log("  Stored saves:");
    for (const [key, save] of this.saves) {
      console.log(`    [${key}] ${save.toString()}`);
    }
  }
}

// --- Main ---

function main(): void {
  const hero = new GameCharacter("Warrior");
  const manager = new SaveManager();

  console.log("=== Starting the adventure ===");
  console.log(hero.status());
  console.log();

  // Play through some actions
  hero.moveTo(10, 20);
  hero.pickUp("Iron Sword");
  hero.levelUp();
  console.log(hero.status());
  console.log();

  // Save at checkpoint 1
  const checkpoint1 = hero.save("Before the dungeon");
  manager.store("checkpoint1", checkpoint1);
  console.log();

  // Enter the dungeon -- things go badly
  hero.moveTo(50, 80);
  hero.takeDamage(60);
  hero.pickUp("Rusty Key");
  hero.takeDamage(30);
  console.log(hero.status());
  console.log();

  // Save the risky state too
  const checkpoint2 = hero.save("Deep in dungeon (low HP)");
  manager.store("checkpoint2", checkpoint2);
  console.log();

  // Even worse...
  hero.takeDamage(50);
  console.log(hero.status());
  console.log();

  // List all saves
  console.log("=== Reviewing saves ===");
  manager.listSaves();
  console.log();

  // Restore to the safe checkpoint
  console.log("=== Restoring to checkpoint 1 ===");
  const save = manager.load("checkpoint1");
  if (save) {
    hero.restore(save);
  }
  console.log(hero.status());
}

main();
